unbricking.md

```diff
@@ -0,0 +1,193 @@
+# G-010S-A Unbricking
+
+If anything goes wrong during an update the SFP module can be unbricked through
+a serial console.
+
+## Preparation / Unlock Bootloader
+<a name="preparation"></a>
+
+By default the bootloader is locked and will not allow interrupting the boot.
+Unlock the bootloader while the SFP module is still working using:
+```
+fw_setenv asc0 0            # Fully activate console
```

Comment:
The "official" way would be to use the script : `serial open`.

This would do :
```
        fw_setenv preboot 'run start_netconsole'
        fw_setenv asc0 0
```
Did you see issues with the `start_netconsole` ?
Also, it seems the default `bootdelay` is already `5` in u-boot (mtd0), but maybe it could have been set to another value.

Reply:
I am fairly certain that bootdelay was 0 on my SFP.

I completely missed that u-boot itself supports netconsole, that acually provides a way easier way without the serial cable. For the "I messed up my boot environment" cases that is actually way easier. The whole serial cable would then be mainly for the "I bricked my SFP without preparing it" case which happens to be the situation I found myself in...

Let me add that to the documentation.

Reply:
done

Reply:
I never managed to get netconsole working.
I get an ARP Request, but my PC doesn't reply.
It seems the link goes down (init by the bootloader, or reset ?), then 1s after it goes up, but it goes down again within 1s.
Increasing bootdelay to 10s doesn't help, maybe the ping doesn't wait enough time and the BCM57810S is too slow.

Reply:
I have the unit in a media converter or switch for most tests, that may change things. And I run it at 1Gbps (switch and media converter do not support 2.5). Let me add a comment for that.

Reply:
Do you see any specific interface parameters when in the bootloader ? Is it 1Gbps full duplex ?

Just to be sure, does `md5sum /dev/mtd0` give `11179aeb378424cb5b18f6e06d2a5667` (ie same uboot as me) ?

I tried changing `ethaddr` (which was set to `ac:9a:96:00:00:00`), but it doesn't help.

Reply:
For both of the SFPs I have:

ONTUSER@SFP:~# md5sum /dev/mtd0
6d655a18f8e6e6211c212cfa0e679fd1  /dev/mtd0

Version info:
U-Boot 2015.04-lantiq-gpon-1.3.6.1 (Feb 07 2018 - 17:12:06)
mips-openwrt-linux-uclibc-gcc (OpenWrt/Linaro GCC 4.8-2014.04 14.07_ltq) 4.8.3
GNU ld (GNU Binutils) 2.22


According to the switch it's connected to it's 1Gbps full duplex

Reply:
I have `U-Boot 2015.04-lantiq-gpon-1.3.6.1 (Jul 11 2017 - 12:55:33)`, so you have a more recent build.
Even if the version is the same, maybe there are fixes.

Reply:
Gut feeling would still be it's caused by having it in a media converter/switch (i.e., does not have to have a fiber inserted to have eth working).

Reply:
With the BCM57810S card, the fiber doesn't need to be present for the Ethernet link to work (so either it doesn't read the RX Loss pin, or the firmware/driver doesn't care).
But maybe there are other requirements, in addition to the "pin 6" mod.

Reply:
Ah ok. I added a somewhat vague comment about netconsole not being consistently working. I assume that if you have a setup where the fiber needs to be connected it will not work (unless the ethernet connection works initially and then gets disabled by one of the drivers if fiber connection is not present).
I also mentioned the exact u-boot version.
I have no pin6 mod active, so that should not be it.  Unless you have it active and it causes an issue, though that would seem weird to me. So I have not mentioned that.
I guess at the moment that is needed so that people don't pull their hair out wondering why netconsole does not work.

For my pure curiosity: If you could make a dump of your u-boot available I'd be happy to try if my netconsole still works with that version.

Reply:
The link between the fiber and the LOS pin is setup in boot scripts in the image.
So in the bootloader, it only depends on the initial LOS pin configuration : it it's in a low state (either as an output, or floating with a pull-down resistor), then it shouldn't prevent the link to be enabled in devices which care about it.

In my case with a the BCM57810S card :
 - It doesn't care about the LOS pin.
 - It requires the pin 6 mod : I have it, currently on the card side.

Here is my u-boot (in gzip, since GitHub doesn't allow to attach bin files) : [uboot-1.3.6.1-20170711.bin.gz](https://github.com/hwti/G-010S-A/files/9897506/uboot-1.3.6.1-20170711.bin.gz)
You can uncompress and flash it from Linux, but we prepared to use your serial cable : if something goes wrong you would need to do the short to recover.

I can try your u-boot too, if you attach it.

Reply:
It's not the specific version, netconsole still works for me:

FALCON => version
version

U-Boot 2015.04-lantiq-gpon-1.3.6.1 (Jul 11 2017 - 12:55:33)
mips-openwrt-linux-uclibc-gcc (OpenWrt/Linaro GCC 4.8-2014.04 14.07_ltq) 4.8.3
GNU ld (GNU Binutils) 2.22
FALCON =>

Reply:
Just in case, here is my uboot version: 
[uboot-1.3.6.1-20180207.bin.gz](https://github.com/hwti/G-010S-A/files/9898101/uboot-1.3.6.1-20180207.bin.gz)

Reply:
Flashing you u-boot didn't change anything.

I tried in a 2.5Gbps media converter (set to 1Gbps by forcing the Ethernet link speed) : the link only goes up at the end of the boot.
I will try the UART in the media converter, if it works maybe it could help to understand why the link doesn't go up.

I tried in a CRS-305 switch : no link, even after waiting enough time (weird).

What is your switch ?
What does `ritool`report as `HardwareVersion` ? `3FE46541AAAA` (like me) or `3FE46541AADA` ?

Reply:
It works for me in a Netgear GS752TP and in a 10Gtek media converter. Both are 1Gbit only.

Hardware version is 3FE46541AADA, so that is one clear difference.

Reply:
I managed to find the TX pin on my media converter, but not the RX pin (probably not connected).

It doesn't matter if it's an initial power on (plugging the SFP) or a reboot, it doesn't wait for the link at all :
```
ROM: V1.1.4
ROM: CFG 0x00000006
ROM: SFLASH-4
Chip:  FALCON-S (A22)
Bootmode: 0x06
Reset cause: Software
CPU Clock: 400 MHz
Net:   SGMII, SERDES [PRIME]
check for netconsole
Error: SGMII TBI not in sync!
Error: SGMII TBI not in sync!
ping failed; host 192.168.1.1 is not alive
Press SPACE to delay and Ctrl-C to abort autoboot in 5 seconds
```

I'm looking at https://github.com/minhng99/u-boot_lantiq_falcon/blob/master/drivers/net/falcon/falcon.c.
The delays are very small, the `Error: SGMII TBI not in sync!` would happen after 2 ms !
Following the source, even if `sgmii_init` fails, `sgmii_aneg` would be called, and it waits up to 200 ms (but the real sources might be different).

In the u-boot variables (ie `fw_printenv`), do you have :
 - `hwconfig` ?
 - `sgmii_inv` ?
 - `ethprime` ? I have `SERDES`, which seems to be the default value.
 - `ethact` ? I have `SERDES` too.
 
But now I get a kernel panic on boot with the media converter, the link never goes up : `BUG: soft lockup - CPU#0 stuck for 22s! [sfp_ddm_tool:1272]`.
The G-010S-A still works fine in the BCM57810S, it's weird...

Reply:
Both eths are SERDES, nothing for the other two.

Here is the full env: 
[env-current.txt.gz](https://github.com/hwti/G-010S-A/files/9914782/env-current.txt.gz)

Reply:
About the Mikrotik CRS305 switch, I just found out it seems to rely on the pin 6 state.
So :
 - With the default (ie `serial close`) configuration : `Module Present` is initially checked, then disables itself during boot, then is enabled again once booted (then the EEPROM values are read, but the links stays down due the Rx Loss).
 - With the `serial open` configuration : `Module Present` is never checked, and nothing works.
 
 I still need to check further with the Rx Loss issue, but it seems we now have a switch on which enabling the UART as a precaution would break the functionality (and require either the pin 6 mod, or to connect the serial adapter to change the configuration in u-boot).

So we probably really need to add a compatibility table instead of the current `Detection and power-up` paragraph, and refer to it as a warning in the unlock instructions.

Reply:
I'll have to try if netconsole works in my CRS305. I would hope that "gpio clear 100; run start_netconsole" enables netconsole while still not causing the pin6 issue. It might come down to timing.

Reply:
I tried with `ethrotate=no`, to make sure u-boot only uses the `ethprime` interface.

On the BCM57810S :
 - `ethprime=SERDES` : ARP request seen by tcpdump, sometimes the link goes down too quickly, sometimes the PC has time to respond, sometimes the netconsole works (works much better on cold boot)
 - `ethprime=SGMII` on cold boot : the link is up, the LED blinks as if there was a transfer, but tcpdump sees nothing, the links stays enabled
 - `ethprime=SGMII` on reboot : the link is down

In the end, `ethrotate=no` doesn't seem to be necessary, but the netconsole timings are very fragile : it often works on cold boot (ie when plugging the SFP, or on reset from u-boot), but almost never on reboot from Linux.
 
Btw, what did you use to pass Ctrl-C to abort the boot ?
I didn't find a netcat which allows it by default, but https://github.com/u-boot/u-boot/blob/master/tools/netconsole works.

Reply:
> I'll have to try if netconsole works in my CRS305. I would hope that "gpio clear 100; run start_netconsole" enables netconsole while still not causing the pin6 issue. It might come down to timing.

There is a `asc0_fixup` command too, but it isn't used in the boot commands.

Reply:
For the compatibility list, I think having a bigger table on the main page would be better : it isn't limited to the unbricking issue, as the pin 6 issue is more general (and we could add a 2.5Gbps column, ...).

Reply:
> In your case, does the netconsole work on reboot too ?
Yeah, I have a long-running nc -u -l -p 6666 and it sees every boot/reboot. I do not remember booting without seeing the console output.
 
> Btw, what did you use to pass Ctrl-C to abort the boot ? I didn't find a netcat which allows it by default, but https://github.com/u-boot/u-boot/blob/master/tools/netconsole works.

I usually use space (caveat is that you can't pause long as it will reboot). You can use socat PTY udp-listen:6666 to link a tty to the netconsole port and then open that pty with a terminal program to get a full console, but that is a lot more involved....

There may be a more specialized netconsole client, but I don't know of one.

Reply:
> There is a `asc0_fixup` command too, but it isn't used in the boot commands.

I think that can only reset to the default values, and I am not 100% sure what it touches, so I stayed away from it.

Reply:
> For the compatibility list, I think having a bigger table on the main page would be better : it isn't limited to the unbricking issue, as the pin 6 issue is more general (and we could add a 2.5Gbps column, ...).

For the netconsole issue I think it belongs here. The other two can be merged and moved to the main page, that makes sense. Have not yet done it yet while we're still collecting data.

Reply:
> Yeah, I have a long-running nc -u -l -p 6666 and it sees every boot/reboot. I do not remember booting without seeing the console output.

OK, so the netconsole (or even networking support in u-boot) can be more or less reliable depending on the the link partner.
Maybe it can be improved, but it's difficult to test when we can't have the UART at the same time.

> I usually use space (caveat is that you can't pause long as it will reboot).

Space didn't seem to work for me.
But what are you able to do with this, without Ctrl-C to interrupt the boot ?

We should probably recommend the u-boot netconsole script.

> For the netconsole issue I think it belongs here.

Maybe, it allows to elaborate too (1Gbps only, reliability, ...).
But It means having two tables, with the pin 6 issue being a common problem.

Reply:
> > Yeah, I have a long-running nc -u -l -p 6666 and it sees every boot/reboot. I do not remember booting without seeing the console output.
> 
> OK, so the netconsole (or even networking support in u-boot) can be more or less reliable depending on the the link partner. Maybe it can be improved, but it's difficult to test when we can't have the UART at the same time.
Reliability may also depend on the hardware version in addition to the link partner.
> Space didn't seem to work for me. But what are you able to do with this, without Ctrl-C to interrupt the boot ?
Sorry, space+enter (I assume some buffering issue..).
I get dropped into the regular console. The difference is that "delay" drops you into console and reboots after seeing no input for a while. Ctrl-C is the same console, but it will not reboot on inactivity.
Bottom line: no long pauses between sf erase and sf write :-)

> 
> We should probably recommend the u-boot netconsole script.
> 
> > For the netconsole issue I think it belongs here.
> 
> Maybe, it allows to elaborate too (1Gbps only, reliability, ...). But It means having two tables, with the pin 6 issue being a common problem.
At the moment I see two tables: One specific for netconsole (e.g., some some hosts not very reliable, maybe also a connection to the hardware version). The other on pin6, requiring fiber, 2.5gig, ...

If we realize that the netconsole issue is independent of hardware, or the pin 6 issue is also connected to the hardware version then the tables may collapse.
On that note: I see the same behavior of "module present" with serial closed and "module not present" with serial open in my Mikrotik. The module does get recognized with pin 100 cleared and netconsole started up. But I'll need some more time setting up the right vlans to test the netconsole. So it would be interesting to see if my hardware version has a pin 6 issue where yours is know to have one.

Reply:
> Sorry, space+enter (I assume some buffering issue..).
> I get dropped into the regular console. The difference is that "delay" drops you into console and reboots after seeing no input for a while. Ctrl-C is the same console, but it will not reboot on inactivity.
> Bottom line: no long pauses between sf erase and sf write :-)

The automatic reboot is very dangerous. 

> So it would be interesting to see if my hardware version has a pin 6 issue where yours is know to have one.

The cause of the pin 6 issue is its use as UART TX (unlike G-010S-P, MA5671A and others, which use another pin). So the only thing which could change is the resistor, which might help for BCM57810S (where `gpio clear 100` + `asc0=3` is not enough for the module to be powered up).

Reply:
Thanks for the instructions that allow me to do additional recovery method. In my setup, I have no problem accessing U-Boot netconsole using the default GNOME terminal in Ubuntu 22.04 and a G-010S-A inserted into an older TP-Link MC220L media converter.
```
# Temporary fix for CTRL+C
# Display all terminal settings
stty -a

# Reassign "Signal Interrupt" to another key
stty intr ^K
```
Use CTRL+C Enter to abort the boot sequence.


I found this info on a Chinese site regarding the purpose of GPIO 100 (translation)

> About preboot for Huawei, Alcatel, Nokia. There is a "gpio input 100", you need to remove (the value is used to check the variables in the uboot_env partition, whether there are changes), otherwise, once the uboot_env variable is changed, the stick will refuse to boot.

Reply:
@tdmadam With your MC220L, does the netconsole work reliably after a `reboot` from Linux ?

Reply:
It reboots fine from linux, I've tried multiple times.

```
FALCON => version
version

U-Boot 2015.04-lantiq-gpon-1.3.6.1 (Feb 07 2018 - 17:12:06)
mips-openwrt-linux-uclibc-gcc (OpenWrt/Linaro GCC 4.8-2014.04 14.07_ltq) 4.8.3
GNU ld (GNU Binutils) 2.22
FALCON => 
```

![netboot1](https://user-images.githubusercontent.com/52431348/199619387-a267a9a0-c2eb-49bc-84a0-ae672945a793.png)

Reply:
Which hardware version do you have?

Reply:
I'm not exactly sure what was there because I overwrote it with another value ```PTIR762-32.7``` needed to match the OMCI value from another ONT.
The only relevant thing that remains and is related to the hardware
```
the HWConfiguration:0000000000000000
the PartNumber:3FE46542AABA
the Variant:BA
```
I am also in possession of a BCM57810S card, I will launch Ubuntu Live later and see how it works with my SFP module.

Reply:
Yeah, PartNumber is what I was looking for. Thanks!

Reply:
> # Temporary fix for CTRL+C
> # Display all terminal settings
> stty -a
> 
> # Reassign "Signal Interrupt" to another key
> stty intr ^K

Ah yes, of course. Thanks!

Reply:
I managed to make the netconsole work on the CRS305 once with :
 - `preboot=gpio clear 100; gpio clear 5; run start_netconsole` : Module Present ON, RX Loss OFF
 - `asc0=0` : it's supposed to be related to the UART, but it seems to disable u-boot console too (with `asc0=3` I get the ARP and ping, but nothing on UDP 6666)

But :
 - It doesn't seem to work on a cold boot : the link is enabled, but nothing seems to be received (Rx Stats at 0).
 - `asc0=0` means Module Present will be OFF once booted to Linux, so the SFP cannot be used.

Reply:
@wteiken The `Compatibility` section should replace `Detection and power-up` , keeping only the additional information about the pin 6 mod (the link, and the warning about the unbricking).

It makes sense to have this at the beginning of the page, since it conditions what can and can't be done with the switch.
I will also add information about mods later, like for the RX Loss.

Reply:
> I managed to make the netconsole work on the CRS305 once with :
> 
>     * `preboot=gpio clear 100; gpio clear 5; run start_netconsole` : Module Present ON, RX Loss OFF
> 
>     * `asc0=0` : it's supposed to be related to the UART, but it seems to disable u-boot console too (with `asc0=3` I get the ARP and ping, but nothing on UDP 6666)
> 
> 
> But :
> 
>     * It doesn't seem to work on a cold boot, unless adding a big delay like 10s : the link is enabled, but nothing seems to be received (Rx Stats at 0).
> 
>     * `asc0=0` means Module Present will be OFF once booted to Linux, so the SFP cannot be used.

There is actually a trick that works for me but that is a bit annoying:
asc0=3 (so no serial input/output)

Use netconsole in client mode:  "nc -u  -p 6666 192.168.1.10 6666"

Run a "ping 192.168.1.10" in a separate terminal.

Once the ping starts responding the NetConsole is active.  Send a few "space+enter" or "ctrl-C+enter" as the ping starts, that actually gives you a netconsole prompt. But you will NOT see the "Press space..." message.

I assume one could also set asc0=0 in preboot, and then back to 3 in bootcmd.

I think the culprit here is /etc/init.d/pin_cfg.sh

A better fix (if creating a patched firmware) would be to patch the section setting up pins in pin_cfg.sh. That is using asc0 to configure the pins at the end of the boot. So patching that read of asc0, of or modifying the uci config for the pins.

Reply:
Patching that script should also be able to permanently disable the LOS signal....

Reply:
I have succeeded in getting the G-010S-A to work with the BCM57810S and display the netconsole on cold start and Linux reboots by going into engineering mode and changing the Net Serdes Interface from SFI to SGMII.

NIC link after inserting SFP module is now almost instantaneous!

Short version:
```ediag_x64.efi -b10eng
device 1
nvm cfg
11 ;(phy)
131=0 ;change from SFI to SGMII
save
exit
```

Reply:
> There is actually a trick that works for me but that is a bit annoying: asc0=3 (so no serial input/output)
> 
> Use netconsole in client mode: "nc -u -p 6666 192.168.1.10 6666"
> 
> Run a "ping 192.168.1.10" in a separate terminal.
> 
> Once the ping starts responding the NetConsole is active. Send a few "space+enter" or "ctrl-C+enter" as the ping starts, that actually gives you a netconsole prompt. But you will NOT see the "Press space..." message.

I didn't know u-boot was listening for an incoming "connection" in this case.

> 
> I assume one could also set asc0=0 in preboot, and then back to 3 in bootcmd.
> 
> I think the culprit here is /etc/init.d/pin_cfg.sh
> 
> A better fix (if creating a patched firmware) would be to patch the section setting up pins in pin_cfg.sh. That is using asc0 to configure the pins at the end of the boot. So patching that read of asc0, of or modifying the uci config for the pins.

In u-boot, having `asc0=0` + `gpio clear 100` seems to work, but I don't know if it would still be the case until the `pin_cfg.sh` (to keep Module Present).

> Patching that script should also be able to permanently disable the LOS signal....

I've done this locally for the CRS305, I will push the mod later since I need to make it configurable first.



> I have succeeded in getting the G-010S-A to work with the BCM57810S and display the netconsole on cold start and Linux reboots by going into engineering mode and changing the Net Serdes Interface from SFI to SGMII.
> 
> NIC link after inserting SFP module is now almost instantaneous!

Do you have `ethact=RGMII` now ?
Did you change `ethprime` ? With `ethprime=SERDES`, `RGMII` is only tried as a second step.
Is it compatible with the `sgmii_mode=5` once the SFP is booted (ie having both 1000 and 2500 working when set with `ethtool`) ?

Reply:
The U-Boot variables ethact=SERDES and ethprime=SERDES remain the same in preboot and full boot mode.
The 2.5Gb mode is only available with the SFI setting.

Reply:
> The U-Boot variables ethact=SERDES and ethprime=SERDES remain the same in preboot and full boot mode. The 2.5Gb mode is only available with the SFI setting.

If `ethact` doesn't change after the `ping ${ncip}`, it means the `SGMII` mode on the BCM57810S still works with the `SERDES` mode in u-boot, and doesn't need to try the `SGMII` mode, which is weird.

Most of the 2.5Gbps patch is in `bnx2x_warpcore_set_sgmii_speed`, which is called by `bnx2x_warpcore_config_sfi` in the `PORT_HW_CFG_NET_SERDES_IF_SFI` case.
It's also called in the `PORT_HW_CFG_NET_SERDES_IF_SGMII` case, but with different parameters : `fiber_mode=0`, `always_autoneg=1` (added by a patch for an external PHY in 2011).
In `bnx2x_populate_int_phy`, `SUPPORTED_2500baseX_Full` has only been added to the `PORT_HW_CFG_NET_SERDES_IF_SFI` case, I guess it would be needed for `PORT_HW_CFG_NET_SERDES_IF_SGMII` too.
But I don't know if `bnx2x_sfp_module_detection => bnx2x_get_edc_mode` would be called.

I wonder if the engineering mode does more than setting static information for the driver.
The difference in the `bnx2x_warpcore_set_sgmii_speed` parameters would be :
 - autonegocation : SGMII mode would force it, SFI mode doesn't allow to enable it (I doubt it would work for 2.5Gbps, but in 1Gbps mode maybe)
 - fibre mode : I don't know what it really means

Reply:
Just to be clear here: NetConsole only needs to work occasionally (when recovering a bricked module). So for the BCM57810S we can now have instructions to get netconsole working,  fix the setup, then change the BCM57810S back to a mode where 2.5gbit is supported if that is desired.

We do not yet have the same set of instructions for a CSR305. For that a patched firmware may be needed, unless going down the path of changing asc0 in preboot and bootcmd. The method I discovered seems a bit too annoying to document for a wider audience.

Or am I misreading something here?

Reply:
@hwti 
> In `bnx2x_populate_int_phy`, `SUPPORTED_2500baseX_Full` has only been added to the `PORT_HW_CFG_NET_SERDES_IF_SFI` case, I guess it would be needed for `PORT_HW_CFG_NET_SERDES_IF_SGMII` too.

If you decide to create a new patch I have a fresh installation with all components ready for testing.
.

@wteiken 
```
ediag_x64.efi -b10eng
device 1
nvm cfg
11 ;(phy)
131=2 ;change back to SFI
save
exit
```

Reply:
> Just to be clear here: NetConsole only needs to work occasionally (when recovering a bricked module). So for the BCM57810S we can now have instructions to get netconsole working, fix the setup, then change the BCM57810S back to a mode where 2.5gbit is supported if that is desired. Actually, the last part I do not have yet. Adam, do you have the instructions to move back to SFI?

Yes, but if we can come up with a better solution, it could both be nice for the development case, and maybe help understanding the issue.

> We do not yet have the same set of instructions for a CSR305. For that a patched firmware may be needed, unless going down the path of changing asc0 in preboot and bootcmd. The method I discovered seems a bit too annoying to document for a wider audience.

Unless I'm doing something wrong, a `setenv asc0 0` in `preboot` isn't enough to get the message on the netconsole.
So the value is probably read earlier, and unless there is a way to make the console fully work without the asc, a custom firmware would be needed (which isn't great, for a new user it would be best to be able to configure the netconsole before trying to flash a firmware).
I also wonder if having `asc0=0` really allows the `gpio clear 100` to work properly, maybe this explains the need for a long delay (`asc0` in u-boot and `onu onu_asc0_pin_cfg_set` in the firmware may configure a pinmux, connecting the pins to the UART or the GPIO controller).

Reply:
> 
> Yes, but if we can come up with a better solution, it could both be nice for the development case, and maybe help understanding the issue.
Ack, just wanted to make sure that I correctly understand that we do have a solution for the moment.

> Unless I'm doing something wrong, a `setenv asc0 0` in `preboot` isn't enough to get the message on the netconsole. So the value is probably read earlier, and unless there is a way to make the console fully work without the asc, a custom firmware would be needed (which isn't great, for a new user it would be best to be able to configure the netconsole before trying to flash a firmware). I also wonder if having `asc0=0` really allows the `gpio clear 100` to work properly, maybe this explains the need for a long delay (`asc0` in u-boot and `onu onu_asc0_pin_cfg_set` in the firmware may configure a pinmux, connecting the pins to the UART or the GPIO controller).
I got my module to respond to a ping in a CSR305 during the u-boot phase, but no netconsole communication yet.  Will have another look tonight. When inserted module is signaling present, neither RX_LOS not TX_FAULT are set. Once boot is fairly far along RX_LOS is triggered as the module is right now not connected to a fiber. I have not yet found a good way to deal with that. I may just connect it via a splitter for more experiments to get the module to communicate when booted. But with the modified pin_cfg at least module present stays on.